include/mrdox/Support/SharedPtr.hpp

```diff
@@ -0,0 +1,325 @@
+//------------------------------------------------
+
+template<class T>
+SharedPtr<T>::
+SharedPtr(
+    Base* p) noexcept
+    : p_(p ? p->addref() : nullptr)
+{
+}
+
+template<class T>
+SharedPtr<T>::
+~SharedPtr()
+{
+    if(p_)
+        p_->release();
+}
+
+template<class T>
+SharedPtr<T>::
+SharedPtr(
+    SharedPtr const& other) noexcept
+    : p_(other.p_ ? other.p_->addref() : nullptr)
+{
+}
+
+template<class T>
+auto
+SharedPtr<T>::
+operator=(
+    SharedPtr const& other) noexcept ->
+    SharedPtr&
+{
+    if(other.p_)
+        other.p_->addref();
+    if(p_)
+        p_->release();
+    p_ = other.p_;
+    return *this;
+}
+
+template<class T>
+template<class U>
+requires std::convertible_to<U*, T*>
+SharedPtr<T>::
+SharedPtr(
+    SharedPtr<U> const& other) noexcept
+    : p_(other.p_ ? other.p_->addref() : nullptr)
+{
+}
+
+template<class T>
+template<class U>
+requires std::convertible_to<U*, T*>
+auto
+SharedPtr<T>::
+operator=(
+    SharedPtr<U> const& other) noexcept ->
+    SharedPtr&
+{
+    if(other.p_)
+        other.p_->addref();
+    if(p_)
+        p_->release();
+    p_ = other.p_;
+    return *this;
+}
+
+template<class T>
+T*
+SharedPtr<T>::
+get() const noexcept
+{
+    return static_cast<T*>(p_->get());
+}
+
+template<class T>
+T*
+SharedPtr<T>::
+operator->() const noexcept
+{
+    return get();
+}
+
+template<class T>
+T&
+SharedPtr<T>::
+operator*() const noexcept
+{
+    return *get();
+}
+
+template<class T>
+SharedPtr<T>::
+operator bool() const noexcept
+{
+    return p_ != nullptr;
+}
+
+template<class T>
+void
+SharedPtr<T>::
+swap(
+    SharedPtr& rhs) noexcept
+{
+    std::swap(p_, rhs.p_);
+}
+
+template<class T>
+template<class U>
+requires std::convertible_to<U*, T*>
+void
+SharedPtr<T>::
+swap(
+    SharedPtr<U>& rhs) noexcept
+{
+    std::swap(p_, rhs.p_);
+}
+
+//------------------------------------------------
+
+template<class U, class... Args>
+SharedPtr<U>
+makeShared(Args&&... args)
+{
+    struct Impl : SharedPtrBase::Base
+    {
+        U u;
+
+        explicit Impl(Args&&... args)
+            : u(std::forward<Args>(args)...)
+        {
+        }
+
+        void* get() noexcept override
+        {
+            return &u;
+        }
+    };
+
+    return SharedPtr<U>(new Impl(
+        std::forward<Args>(args)...));
+}
+
+//------------------------------------------------
+
+template<class T>
+class AtomicSharedPtr : protected SharedPtrBase
+{
+    using SharedPtrBase::Base;
+
+    std::atomic<Base*> pp_ = nullptr;
+
+public:
+    SharedPtr<T>
+    load() const noexcept
+    {
+        return SharedPtr<T>(pp_.load());
+    }
+
+    /** Return an existing or newly constructed object with shared ownership.
+
+        InitFunction must have this equivalent signature:
+        @code
+        SharedPtr<T>(void)
+        @endcode
+    */
+    template<class InitFunction>
+    requires std::is_invocable_r_v<
+        SharedPtr<T>, InitFunction>
+    SharedPtr<T>
+    load(InitFunction const& init)
```

Comment:
Might be overkill, but I'd take this function as a `InitFunction&&`, then use it as `std::invoke(std::forward<InitFunction>(init))`

Reply:
Hmm interesting what is the difference?

Reply:
The whole incantation with `invoke` probably doesn't matter, since it has to be invocable with no arguments anyway. `init()` should be fine. Personal preference, I just like the look of the uniform invocation with `invoke`.

I'd still use a forwarding reference for the function type, otherwise any rvalue would get stored in this lvalue-const-ref, and someone looking at this later might do something with it they shouldn't

Reply:
"otherwise any rvalue would get stored in this lvalue-const-ref"

No idea what this means, I don't speak nerd

Reply:
You might get weird errors one day if someone makes an unexpected change. For now it's fine

Reply:
Can you give an example of how this code my be invoked which could cause a problem? That would help me visualize it.